File: yarvis-IA/parser_py/parser_csv.py
"""
Parser de catálogos en formato CSV (.csv, .tsv).
Detecta separadores automáticamente (, o ;).
"""
from core.utils import limpiar_producto


def _detectar_separador_csv(linea: str) -> str | None:
    """Detecta si una línea es CSV y retorna el separador (, o ;)."""
    comas = linea.count(',')
    puntos_coma = linea.count(';')
    
    if comas >= 1 and comas > puntos_coma:
        return ','
    elif puntos_coma >= 1 and puntos_coma > comas:
        return ';'
    return None


def parsear_csv(texto: str) -> list[dict]:
    """
    Parsea un catálogo en formato CSV.
    
    Formatos soportados:
        nombre,costo,venta,categoria,stock
        nombre;costo;venta;categoria;stock
        nombre,venta,costo  (detecta automáticamente el orden)
    """
    lineas = texto.strip().splitlines()
    if not lineas:
        return []
    
    # Detectar separador de la primera línea
    separador = _detectar_separador_csv(lineas[0])
    if not separador:
        return []
    
    # Detectar si la primera línea es header
    primera_linea = lineas[0].lower()
    tiene_header = any(p in primera_linea for p in ['nombre', 'producto', 'precio', 'costo', 'venta', 'categoria', 'stock', 'cantidad', 'existencia'])
    
    # Detectar headers si existen
    headers = []
    if tiene_header:
        headers = [h.strip().lower() for h in lineas[0].split(separador)]
    
    # Buscar índice de columna de stock/cantidad
    col_stock = None
    for i, h in enumerate(headers):
        if any(k in h for k in ['stock', 'existencia', 'cantidad', 'inventario', 'qty', 'quantity']):
            col_stock = i
            break
    
    # Buscar índice de columna de categoría
    col_categoria = None
    for i, h in enumerate(headers):
        if any(k in h for k in ['categoría', 'categoria', 'tipo', 'seccion', 'sección', 'departamento', 'category', 'type']):
            col_categoria = i
            break
    
    productos = []
    
    for i, linea in enumerate(lineas):
        linea = linea.strip()
        if not linea:
            continue
        
        # Saltar header si existe
        if tiene_header and i == 0:
            continue
        
        partes = [p.strip().strip('"').strip("'") for p in linea.split(separador)]
        
        # Necesitamos al menos 2 columnas (nombre + precio)
        if len(partes) < 2:
            continue
        
        # Intentar detectar el orden de columnas
        nombre = None
        precio_venta = None
        precio_costo = None
        categoria = None
        stock = 0
        
        # Buscar columnas numéricas
        numeric_cols = []
        text_cols = []
        for j, p in enumerate(partes):
            p_clean = p.replace('$', '').replace(',', '').strip()
            try:
                val = float(p_clean)
                numeric_cols.append((j, val))
            except ValueError:
                if p_clean:
                    text_cols.append((j, p_clean))
        
        # La primera columna de texto es el nombre
        for j, p in text_cols:
            if not p.replace('.', '').isdigit():
                nombre = p
                break
        
        if not nombre:
            continue
        
        # Si solo hay 2 columnas de texto y no hay precios, 
        # la primera puede ser categoría y la segunda producto
        if len(numeric_cols) == 0 and len(text_cols) == 2:
            first_text = text_cols[0][1]
            second_text = text_cols[1][1]
            # El nombre más largo probablemente es el producto
            if len(second_text) > len(first_text):
                nombre = second_text
                categoria = first_text.upper()
            else:
                nombre = first_text
                categoria = second_text.upper()
        
        # Asignar precios
        if len(numeric_cols) >= 2:
            idx1, val1 = numeric_cols[0]
            idx2, val2 = numeric_cols[1]
            
            if val1 > val2:
                precio_venta = val1
                precio_costo = val2
            else:
                precio_venta = val2
                precio_costo = val1
            
            # Buscar categoría en la última columna de texto
            for j, p in reversed(text_cols):
                if p != nombre and len(p) < 30:
                    categoria = p.upper()
                    break
        elif len(numeric_cols) == 1:
            _, val = numeric_cols[0]
            precio_venta = val
            precio_costo = 0
        
        # Obtener stock de la columna detectada
        if col_stock is not None and col_stock < len(partes):
            stock_str = partes[col_stock].replace(',', '').strip()
            try:
                stock = int(float(stock_str))
            except (ValueError, TypeError):
                stock = 0
        
        # Obtener categoría de la columna detectada
        if col_categoria is not None and col_categoria < len(partes) and partes[col_categoria]:
            categoria = partes[col_categoria].strip().upper()
        
        # Si no hay categoría aún, buscar en columnas de texto que no sean el nombre
        if categoria == "SIN CATEGORÍA" or categoria is None:
            for j, p in text_cols:
                if p != nombre and len(p) < 30:
                    categoria = p.upper()
                    break
        
        # Si solo hay 2 columnas de texto, la primera puede ser categoría
        if (categoria == "SIN CATEGORÍA" or categoria is None) and len(text_cols) == 2:
            first_text = text_cols[0][1]
            second_text = text_cols[1][1]
            if first_text != nombre:
                categoria = first_text.upper()
            elif second_text != nombre:
                categoria = second_text.upper()
        
        if precio_venta and precio_venta > 0:
            productos.append({
                "nombre": limpiar_producto(nombre),
                "precio_costo": round(precio_costo or 0, 2),
                "precio_venta": round(precio_venta, 2),
                "stock": stock,
                "categoria": categoria or "SIN CATEGORÍA",
            })
        elif nombre:
            # Sin precios, agregar con precio 0
            productos.append({
                "nombre": limpiar_producto(nombre),
                "precio_costo": 0,
                "precio_venta": 0,
                "stock": stock,
                "categoria": categoria or "SIN CATEGORÍA",
            })
    
    return productos

File: yarvis-app/src-tauri/src/parser_rs/parser_csv.rs
// Parser CSV en Rust (local, sin sidecar).
// Lee archivos .csv y .tsv directamente.
use std::fs;
use crate::models::InventoryItem;
use super::utils::sanitize_path;

// ============================================================
// Parser CSV
// ============================================================

/// Detecta si una línea es CSV y retorna el separador
fn detectar_separador(line: &str) -> Option<char> {
    let comas = line.matches(',').count();
    let puntos_coma = line.matches(';').count();

    if comas >= 2 && comas > puntos_coma {
        Some(',')
    } else if puntos_coma >= 2 && puntos_coma > comas {
        Some(';')
    } else {
        None
    }
}

#[tauri::command]
pub fn parsear_catalogo_csv(path: String) -> Result<Vec<InventoryItem>, String> {
    let safe_path = sanitize_path(&path)?;
    let content = fs::read_to_string(safe_path).map_err(|e| e.to_string())?;
    let mut items = Vec::new();

    let lines: Vec<&str> = content.lines().collect();
    if lines.is_empty() {
        return Ok(items);
    }

    // Detectar separador
    let separador = detectar_separador(lines[0])
        .ok_or("No se pudo detectar separador CSV (use , o ;)")?;

    // Detectar header
    let primera_linea = lines[0].to_lowercase();
    let tiene_header = ["nombre", "producto", "precio", "costo", "venta", "categoria"]
        .iter()
        .any(|p| primera_linea.contains(p));

    let data_start = if tiene_header { 1 } else { 0 };

    for line in &lines[data_start..] {
        let line_trimmed = line.trim();
        if line_trimmed.is_empty() {
            continue;
        }

        let parts: Vec<&str> = if separador == ',' {
            line_trimmed.split(',').collect()
        } else {
            line_trimmed.split(';').collect()
        };

        if parts.len() < 2 {
            continue;
        }

        // Buscar nombre (primera columna no numérica)
        let mut nombre = String::new();
        let mut precio_venta = 0.0;
        let mut precio_costo = 0.0;
        let mut numeric_vals: Vec<(usize, f64)> = Vec::new();

        for (i, part) in parts.iter().enumerate() {
            let clean = part.replace('$', "").replace(',', "").trim().to_string();
            if let Ok(val) = clean.parse::<f64>() {
                numeric_vals.push((i, val));
            } else if nombre.is_empty() && !clean.is_empty() {
                nombre = clean.to_uppercase();
            }
        }

        if nombre.is_empty() || numeric_vals.is_empty() {
            continue;
        }

        // Asignar precios
        if numeric_vals.len() >= 2 {
            let (_, val1) = numeric_vals[0];
            let (_, val2) = numeric_vals[1];
            if val1 > val2 {
                precio_venta = val1;
                precio_costo = val2;
            } else {
                precio_venta = val2;
                precio_costo = val1;
            }
        } else if numeric_vals.len() == 1 {
            precio_venta = numeric_vals[0].1;
        }

        if precio_venta > 0.0 {
            items.push(InventoryItem {
                id: None,
                nombre,
                descripcion: None,
                precio_costo,
                precio_venta,
                stock: 0.0,
                stock_minimo: 5.0,
                vendido: 0.0,
                codigo_barras: None,
                categoria: None,
            });
        }
    }

    Ok(items)
}
